firestore/src/jni/arena_ref.cc

```diff
@@ -0,0 +1,146 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "firestore/src/jni/arena_ref.h"
+
+#include <atomic>
+#include <utility>
+
+#include "firestore/src/android/firestore_android.h"
+#include "firestore/src/jni/env.h"
+#include "firestore/src/jni/hash_map.h"
+#include "firestore/src/jni/long.h"
+#include "firestore/src/jni/map.h"
+
+namespace firebase {
+namespace firestore {
+namespace jni {
+
+namespace {
+
+HashMap* gArenaRefHashMap = nullptr;
+jclass gLongClass = nullptr;
+jmethodID gLongConstructor = nullptr;
+std::mutex mutex_;
+
+int64_t GetNextArenaRefKey() {
+  static std::atomic<int64_t> next_key(0);
+  return next_key.fetch_add(1);
+}
+
+}  // namespace
+
+ArenaRef::ArenaRef(Env& env, const Object& object)
+    : key_(GetNextArenaRefKey()) {
+  Local<Long> key_ref = key_object(env);
+  std::lock_guard<std::mutex> lock(mutex_);
+  gArenaRefHashMap->Put(env, key_ref, object);
+}
+
+ArenaRef::ArenaRef(const ArenaRef& other)
+    : key_(other.key_ == kInvalidKey ? kInvalidKey : GetNextArenaRefKey()) {
+  if (other.key_ != kInvalidKey) {
+    Env env;
+    Local<Object> object = other.get(env);
+    Local<Long> key_ref = key_object(env);
+    std::lock_guard<std::mutex> lock(mutex_);
+    gArenaRefHashMap->Put(env, key_ref, object);
+  }
+}
+
+ArenaRef::ArenaRef(ArenaRef&& other) {
+  using std::swap;  // go/using-std-swap
+  swap(key_, other.key_);
+}
+
+ArenaRef& ArenaRef::operator=(const ArenaRef& other) {
+  if (this == &other || (key_ == kInvalidKey && other.key_ == kInvalidKey)) {
+    return *this;
+  }
+
+  Env env;
+  if (other.key_ != kInvalidKey) {
+    if (key_ == kInvalidKey) {
+      key_ = GetNextArenaRefKey();
+    }
+    Local<Object> object = other.get(env);
+    std::lock_guard<std::mutex> lock(mutex_);
+    gArenaRefHashMap->Put(env, key_object(env), object);
+  } else {
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      gArenaRefHashMap->Remove(env, key_object(env));
+    }
+    key_ = kInvalidKey;
+  }
+  return *this;
+}
+
+ArenaRef& ArenaRef::operator=(ArenaRef&& other) {
+  if (this != &other) {
+    std::swap(key_, other.key_);
+  }
+  return *this;
+}
+
+ArenaRef::~ArenaRef() {
+  if (key_ != kInvalidKey) {
+    Env env;
+    ExceptionClearGuard block(env);
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      gArenaRefHashMap->Remove(env, key_object(env));
+    }
+    if (!env.ok()) {
+      env.RecordException();
+      env.ExceptionClear();
+    }
+  }
+}
+
+Local<Long> ArenaRef::key_object(Env& env) const {
+  if (!env.ok()) return {};
+  // The reason why using raw JNI calls here rather than the JNI convenience
+  // layer is because we need to get rid of JNI convenience layer dependencies
+  // in the ArenaRef destructor to avoid racing condition when firestore object
+  // gets destroyed.
+  jobject key = env.get()->NewObject(gLongClass, gLongConstructor, key_);
+  env.RecordException();
```

Comment:
Calling `RecordException()` isn't sufficient to guard against exceptions. Instead, check `env.ok()` and return `{}` if it's not ok.